### ISSUE TITLE
Mark all checks with appropriate annotation so that they could be used in the MT mode

### ISSUE THREAD

**Issue by soon** [checkstyle/checkstyle#4870]
Subtask of https://github.com/checkstyle/checkstyle/issues/4473

Each check must implement either `OneCheckInstancePerApplication` or `OneCheckInstancePerThread`.

All changes for this task must be just `implements OneCheckInstancePerApplication` or `implements OneCheckInstancePerThread`, i.e. no more logic is expected. If a particular check requires additional changes, there should be a separate issue for that check.

-----------

> Please expand in issue what are the requirements for a module to be marked with a specific interface.

1, If a check holds file state (i.e. non-final fields, non-property fields) it must be marked with `OneCheckInstancePerThread` interface.
2. If a check is stateless (all fields are final or properties, and there are no mutable fields like sets or lists) it must be marked with `OneCheckInstancePerApplication`. Note, that  a "mutable field" here - is the field which may change it's state while the Check processes a file.
3. If a check is global-stateful (it contains a mutable state, which must be updated across the whole application lifetime), it must be marked with `OneCheckInstancePerApplication` and all state updates must be thread-safe.

**Comment by rnveach** [comment]
> If a check is a global-stateful 

Why don't we have a unique interface for this?
Even if it acts the same as `OneCheckInstancePerApplication` won't there be some benefit from separating them out even if it is just for users/us to know?

> If a check holds file state (i.e. non-final fields, non-property fields) it must be marked with OneCheckInstancePerThread interface.

By this definition alone, doesn't it mean `AbstractCheck` holds state as it has a few non-final fields. Whatever we define `AbstractCheck` as, it will extend to all classes that extend it too.
Or do we only examine change in states after it has been initialized/cloned?

**Comment by soon** [comment]
> Why don't we have a unique interface for this?
> Even if it acts the same as OneCheckInstancePerApplication won't there be some benefit from separating them out even if it is just for users/us to know?

@rnveach  We can, sure, but I don't see a reason for it. Maybe in the future it will be useful. Should I add it?

> By this definition alone, doesn't it mean AbstractCheck holds state as it has a few non-final fields. Whatever we define AbstractCheck as, it will extend to all classes that extend it too.
> Or do we only examine change in states after it has been initialized/cloned?

The `AbstractCheck` is global stateful now, yes. It will require some changes, so that `messages` could be used from multiple threads.

**Comment by rnveach** [comment]
> Should I add it?

I think we should.
@romani What do you think?

> It will require some changes, so that messages could be used from multiple threads.

Doesn't this have to be done first before we even implement the MT modes?

> The AbstractCheck is global stateful now, yes

If `messages` is the problem, isn't it more file stateful then as messages are based on a file and not the entire application as a whole? 
Also if `messages` is the problem, it is declared `final` and is `stateless` based on your written requirements.

**Comment by soon** [comment]
> Also if messages is the problem, it is declared final and is stateless based on your written requirements.

I've updated the post

> Doesn't this have to be done first before we even implement the MT modes?

Created issue https://github.com/checkstyle/checkstyle/issues/4908, preparing PR

> If messages is the problem, isn't it more file stateful then as messages are based on a file and not the entire application as a whole?

Not sure if I understood you properly. We cannot make the `AbstractCheck` file-stateful, because all subclassess will be file-statefull too, no mater if they contain a state or not.

**Comment by rnveach** [comment]
> We cannot make the AbstractCheck file-stateful, because all subclassess will be file-statefull too, no mater if they contain a state or not.

That is correct, and what I was talking about in one of the PRs where you were marking some `abstract` classes.

My statement was how can `AbstractCheck` be completely stateless if it holds messages of a file? That alone makes it sound like it will never be anything but file-stateful until messages it completely removed that class.
I haven't looked at your new PR yet.

**Comment by soon** [comment]
I want to discuss the [`IllegalTypeCheck`](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java). I marked it as global-stateful check, because it updates field `illegalClassNames` across the application lifetime. In this case we have to make this field thread-safe. 
However, seems like the check could be file-stateful as well

@rnveach @sabaka @romani  what do you think?

**Comment by rnveach** [comment]
> seems like the check could be file-stateful as well

It should be file-stateful as it has to look at imports for a class, imo.

>  it updates field illegalClassNames across the application lifetime.

I assume you are referring to this line: https://github.com/checkstyle/checkstyle/blob/78de3dc771531cfee50a60ff2cee80c46b83c43b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java#L352 as it is called in `visitImport`.
This does seem like weird behavior. What is even more weird, if I remove all the code in `visitImport` no tests fail. Mutation is a 95%, so this could be the class that is holding it back. I would need to run regression with it gone to see if there is some behavior we are missing.

Feel free to create a new issue on this.

**Comment by rnveach** [comment]
@soon I am moving issue with IllegalTypeCheck to a new issue.

**Comment by romani** [comment]
@soon , is it ok to close this issue ?

**Comment by rnveach** [comment]
@romani I made sure un-annotated checks are treated no differently then they are today.
See also my comment at https://github.com/checkstyle/checkstyle/issues/5140#issuecomment-331947516 re-iterating this. We placed it into documentation at
https://github.com/checkstyle/checkstyle/blob/ebca211ef9f8a757a8973cc7e3252c80c736f9ca/src/main/java/com/puppycrawl/tools/checkstyle/FileStatefulCheck.java#L43-L44

**Comment by rnveach** [comment]
I am not closing this issue yet.
I think we should add a UT to show all checks are marked. We added some checks since this issue was originally opened.

**Comment by rnveach** [comment]
I will work on the test.

**Comment by romani** [comment]
Fix is merged
